Fix ProcessSwap.ReuseProcessAfterNavigatingFromAboutPage API test failure in site isolation

colelao opened the pull request:
#### 58362ea9823932fea4c9751d1487a5e226bfc5ef
<pre>
Fix ProcessSwap.ReuseProcessAfterNavigatingFromAboutPage API test failure in site isolation
<a href="https://bugs.webkit.org/show_bug.cgi?id=300273">https://bugs.webkit.org/show_bug.cgi?id=300273</a>
<a href="https://rdar.apple.com/162078528">rdar://162078528</a>

Reviewed by Sihui Liu.

This patch adds a case to WebProcessPool::processForNavigationInternal.
When the protocol is about, navigations were treated as cross-site. This patch
adds a check for site isolation to fix the bug.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didStartUsingProcessForSiteIsolation):

Canonical link: <a href="https://commits.webkit.org/301488@main">https://commits.webkit.org/301488@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d13970aaed9590f90b1ec5cc1a031d93edfe140

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125955 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45615 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36385 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132820 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77810 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5a6da261-b924-452d-8887-10cf8a613465) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46298 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54172 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95960 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64054 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a699033a-7bb1-4438-b830-ad2e33b8461b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128903 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37031 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112647 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76533 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b6e5a4ce-667a-4e29-8446-edb868e002f8) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35939 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76291 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106813 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31046 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135507 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52737 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40466 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104431 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53186 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108859 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104151 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26569 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49530 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27861 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50115 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52631 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58446 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51975 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55323 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53672 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2350,9 +2350,21 @@ std::tuple<Ref<WebProcessProxy>, RefPtr<SuspendedPageProxy>, ASCIILiteral> WebPr
+            if (sourceProcess->hasCommittedAnyMeaningfulProvisionalLoads())
+                return false;
+            if (siteIsolationEnabled) {
+                if (sourceProcess->site()->isEmpty())
```
I think you need to null check `sourceProcess->site()` before invoking `isEmpty()` on it.

szewai commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2350,9 +2350,21 @@ std::tuple<Ref<WebProcessProxy>, RefPtr<SuspendedPageProxy>, ASCIILiteral> WebPr
+                return false;
+            if (siteIsolationEnabled) {
+                if (sourceProcess->site()->isEmpty())
+                    return sourceProcess->isSharedProcess();
```
I think this check is a bit confusing:
`isSharedProcess()` returns true if `!sourceProcess->site() && sourceProcess->site().error() == SiteState::SharedProcess`, so if process has an empty m_site (`sourceProcess->site()->isEmpty()`), this check will never return true?

szewai commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2350,9 +2350,23 @@ std::tuple<Ref<WebProcessProxy>, RefPtr<SuspendedPageProxy>, ASCIILiteral> WebPr
         }
+
         // With site isolation enabled, this condition is not enough to indicate the web process can be reused;
         // we may also need to consider whether the process is used or in use by other sites.
```
The comment might be removed.

colelao committed:
Fix ProcessSwap.ReuseProcessAfterNavigatingFromAboutPage API test failure in site isolation
https://bugs.webkit.org/show_bug.cgi?id=300273
rdar://162078528

Reviewed by Sihui Liu.

This patch adds a case to WebProcessPool::processForNavigationInternal.
When the protocol is about, navigations were treated as cross-site. This patch
adds a check for site isolation to fix the bug.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didStartUsingProcessForSiteIsolation):

Canonical link: https://commits.webkit.org/301488@main
